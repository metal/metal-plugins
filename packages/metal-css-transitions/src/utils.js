'use strict';

import {isFunction, isString} from 'metal';

const NAMESPACE = '__METAL_CSS_TRANSISIONS__';

/**
 * Processes an array of components, creating an object with keys and values.
 * If the component does not already have a key, it will be given a unique key.
 * @param {!Array} children Array of metal components.
 */
export function getChildrenMap(children) {
	let COUNTER = 1;

	const retMap = new Map();

	const keys = children.map(
		child => (child.props ? child.props.key : undefined)
	);

	children.forEach(child => {
		if (child && child.props) {
			let {props: {key}, tag} = child;

			if (!key) {
				if (!tag) {
					tag = NAMESPACE;
				} else if (!isString(tag)) {
					tag = isFunction(tag) ? tag.name : NAMESPACE;
				}

				while (keys.indexOf(`${tag}-${COUNTER}`) !== -1) {
					COUNTER += 1;
				}

				key = `${tag}-${COUNTER}`;

				keys.push(key);

				child.props.key = key;
			}

			retMap.set(key, child);
		}
	});

	return retMap;
}

/**
 * Merges two children maps so that there are no duplicates.
 * @param {!Object} next Map of new children components.
 * @param {!Object} prev Map of previous children components.
 */
const map = new Map();

export function mergeChildrenMap(nextMap = map, prevMap = map) {
	function getValueForKey(key) {
		if (nextMap.has(key)) {
			return nextMap.get(key);
		} else {
			return prevMap.get(key);
		}
	}

	const nextKeysPending = {};

	let pendingKeys = [];

	prevMap.forEach((value, prevKey) => {
		if (nextMap.has(prevKey)) {
			if (pendingKeys.length) {
				nextKeysPending[prevKey] = pendingKeys;

				pendingKeys = [];
			}
		} else {
			pendingKeys.push(prevKey);
		}
	});

	const mergedMap = new Map();

	for (let i = 0; i < pendingKeys.length; i++) {
		mergedMap.set(pendingKeys[i], getValueForKey(pendingKeys[i]));
	}

	nextMap.forEach((value, nextKey) => {
		if (nextKeysPending.hasOwnProperty(nextKey)) {
			for (let i = 0; i < nextKeysPending[nextKey].length; i++) {
				const pendingNextKey = nextKeysPending[nextKey][i];

				mergedMap.set(
					nextKeysPending[nextKey][i],
					getValueForKey(pendingNextKey)
				);
			}
		}

		mergedMap.set(nextKey, getValueForKey(nextKey));
	});

	return mergedMap;
}
