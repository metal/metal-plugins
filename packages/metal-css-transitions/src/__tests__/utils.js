'use strict';

import {getChildrenMap, mergeChildrenMap} from '../utils';

describe('getChildrenMap', () => {
	it('should create a keyed object based on props.key', () => {
		const arr = [
			{
				props: {
					key: 'foo'
				},
				tag: 'div'
			},
			{
				props: {
					key: 'bar'
				},
				tag: 'div'
			}
		];

		const obj = getChildrenMap(arr);

		expect(obj).toMatchSnapshot();
	});

	it('should create a keyed object based on props.key when keys are numbers', () => {
		const arr = [
			{
				props: {
					key: 2
				},
				tag: 'div'
			},
			{
				props: {
					key: 1
				},
				tag: 'div'
			}
		];

		const obj = getChildrenMap(arr);

		expect(obj).toMatchSnapshot();
	});

	it('should create a keyed object even if tag is all sorts of types', () => {
		const arr = [
			{
				props: {},
				tag: 'div'
			},
			{
				props: {},
				tag: function testTest() {}
			},
			{
				props: {},
				tag: {foo: 'bar'}
			},
			{
				props: {}
			}
		];

		const obj = getChildrenMap(arr);

		expect(obj).toMatchSnapshot();
	});
});

describe('mergeChildrenMap', () => {
	it('should merge objects', () => {
		const newMap = new Map([[1, 'one'], [2, 'two']]);

		const oldMap = new Map([[2, 'twotwo'], [3, 'three']]);

		let mergedMap = mergeChildrenMap(newMap, oldMap);

		expect(mergedMap.get(2)).toBe('two');

		mergedMap = mergeChildrenMap(oldMap, newMap);

		expect(mergedMap.get(2)).toBe('twotwo');
	});

	it('should still work if no objects are supplied', () => {
		const obj = mergeChildrenMap();

		expect(obj instanceof Map);
	});
});
