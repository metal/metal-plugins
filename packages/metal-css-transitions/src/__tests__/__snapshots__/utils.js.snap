// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getChildrenMap should create a keyed object based on props.key 1`] = `
Map {
  "foo" => Object {
    "props": Object {
      "key": "foo",
    },
    "tag": "div",
  },
  "bar" => Object {
    "props": Object {
      "key": "bar",
    },
    "tag": "div",
  },
}
`;

exports[`getChildrenMap should create a keyed object based on props.key when keys are numbers 1`] = `
Map {
  2 => Object {
    "props": Object {
      "key": 2,
    },
    "tag": "div",
  },
  1 => Object {
    "props": Object {
      "key": 1,
    },
    "tag": "div",
  },
}
`;

exports[`getChildrenMap should create a keyed object even if tag is all sorts of types 1`] = `
Map {
  "div-1" => Object {
    "props": Object {
      "key": "div-1",
    },
    "tag": "div",
  },
  "testTest-1" => Object {
    "props": Object {
      "key": "testTest-1",
    },
    "tag": [Function],
  },
  "__METAL_CSS_TRANSISIONS__-1" => Object {
    "props": Object {
      "key": "__METAL_CSS_TRANSISIONS__-1",
    },
    "tag": Object {
      "foo": "bar",
    },
  },
  "__METAL_CSS_TRANSISIONS__-2" => Object {
    "props": Object {
      "key": "__METAL_CSS_TRANSISIONS__-2",
    },
  },
}
`;
